### PR TITLE
use SChannel for git during get source.

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 1]
```diff
@@ -424,6 +424,21 @@ public AgentSettings LoadSettings()
 
             _term.WriteLine(StringUtil.Loc("SavedSettings", DateTime.UtcNow));
 
+            Boolean saveRuntimeOptions = false;
+            AgentRuntimeOptions runtimeOptions = new AgentRuntimeOptions();
```
Comment (ericsciple) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: var

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 2]
```diff
@@ -563,6 +578,9 @@ public AgentSettings LoadSettings()
                     // delete agent cert setting
                     (HostContext.GetService<IAgentCertificateManager>() as AgentCertificateManager).DeleteCertificateSetting();
 
+                    // delete agent runtime option
+                    _store.DeleteAgentRuntimeOptions();
```
Comment (ericsciple) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: should this be best effort? is it different than deleting the rest, or should it just be consistent?

File: src/Agent.Worker/Build/GitSourceProvider.cs [hunk 3]
```diff
@@ -56,7 +63,14 @@ public override bool GitLfsUseAuthHeaderCmdlineArg
 
         public override void RequirementCheck(IExecutionContext executionContext, ServiceEndpoint endpoint)
         {
-            // no-opt, there is no additional requirements.
+#if OS_WINDOWS
+            // check git version for SChannel SSLBackend (Windows Only)
+            bool schannelSslBackend = HostContext.GetService<IConfigurationStore>().GetAgentRuntimeOptions()?.GitUseSecureChannel ?? false;
+            if (schannelSslBackend)
+            {
+                _gitCommandManager.EnsureGitVersion(_minGitVersionSupportSSLBackendOverride, throwOnNotMatch: true);
```
Comment (ericsciple) [inline src/Agent.Worker/Build/GitSourceProvider.cs]: you could also skip this if not system.prefergit

Comment (ericsciple) [reply src/Agent.Worker/Build/GitSourceProvider.cs]: nm, result is cached

File: src/Microsoft.VisualStudio.Services.Agent/ConfigurationStore.cs [hunk 4]
```diff
@@ -284,5 +302,35 @@ public void DeleteAutoLogonSettings()
...
+            if (File.Exists(_runtimeOptionsFilePath))
+            {
+                // Delete existing runtime options file first, since the file is hidden and not able to overwrite.
+                Trace.Info("Delete exist runtime options file.");
+                IOUtil.DeleteFile(_runtimeOptionsFilePath);
+            }
+
+            IOUtil.SaveObject(options, _runtimeOptionsFilePath);
+            Trace.Info("Options Saved.");
+            File.SetAttributes(_runtimeOptionsFilePath, File.GetAttributes(_runtimeOptionsFilePath) | FileAttributes.Hidden);
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/ConfigurationStore.cs]: hidden? are all others? proxy?

File: src/Microsoft.VisualStudio.Services.Agent/Constants.cs [hunk 5]
```diff
@@ -317,6 +318,7 @@ public static class Features
                 public static readonly string BuildDirectoryClean = "agent.clean.buildDirectory";
                 public static readonly string GitLfsSupport = "agent.source.git.lfs";
                 public static readonly string GitShallowDepth = "agent.source.git.shallowFetchDepth";
+                public static readonly string SChannelSslBackend = "agent.source.git.schannel";
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/Constants.cs]: ?

**Comment by ericsciple** [comment]
makes sense

**Comment by TingluoHuang** [comment]
@scottboehmer FYI, i am adding option to support using SecureChannel in git source provide

**Comment by mingalsuo** [comment]
@TingluoHuang 

Should this just turn on by itself? We just **updated** to v2.129.0 and still see the 'SSL certificate problem: unable to get local issuer certificate' -message even though the machine trusts the CA in question, proven by running an external instance of Git with SChannel enabled.

**Comment by bryanmacfarlane** [comment]
@mingalsuo - no, it's a config time option (gituseschannel)

@TingluoHuang - is this in the docs yet?

**Comment by mingalsuo** [comment]
Seems to be documented, thanks!

https://docs.microsoft.com/en-us/vsts/build-release/actions/agents/certificate

**Pull request merged**